CHANGELOG.md

```diff
@@ -1,5 +1,13 @@
 _The versioning refers to the React component build._
 
+#### v3.0.0 (2018-03-09)
+* React: remove obsolete @ssr-ready pragma from components.
+* React: allow importing individual icons, both in CommonJS and ECMAScript module formats.
+* React: substitute React.PureComponent for a JavaScript function, to minimize the bundle size of transpiled CommonJS files as much as possible.
+* Package: added a `esnext` key in the package.json so API consumers can use it for importing the main ESM file.
+* Build: use template literals to create the React components and centralize the _icon needs offset_ logic in the svg-to-react Grunt task.
+* Build: deleted the `build/` directory in favor of separate `cjs/` and `esm/` folders that contain CommonJS and ECMAScript modules respectively.
+
```

Comment:
Can we avoid acronyms (cjs, esm) in favor of a clear, spelled out folder name?

* `common-js`
* `ecmascript`

Also: can we add somewhere a detail on why there's this split, and as a developer when should I pick one or the other?

In genereal: is there any way to avoid this split for a simpler solution that reduces a choice requirement?

Reply:
That's a very good question I myself pondered about. 

We can have a default `import GridiconExternal from 'gridicons/external'` that pulls the CommonJS module (which is now the common format used in npm packages). That was actually the original proposal, but it would compromise future backward compatibility, as it was raised by @ockham Some notes on this:

* We live in times of change when it comes to JavaScript module formats. Many libraries are shipping both `cjs` and `esm`. There is no a strong standard on how to do it yet. I wish we could just avoid this module war, but we can't.
* If we publish now the `cjs` icons under `gridicons`, when `esm` becomes more used we won't be able to use that namespace and maintain backward compatibility. We'll need to publish the individual icons under `gridicons/esm`, for example. This would put us in a weird position and could potentially introduce errors if the consumer is expecting a `esm` module when using `gridicons/external`, for example.
* In the README, the `esm` option is deemphasized, which was intentional to mitigate decision fatigue.
* Consumers of the API can always teach their module bundlers to make `gridicons` an alias for `gridicons/cjs` or `gridicons/esm` - this is probably what we'll do in Calypso.

Does this address your questions @folletto ?

Reply:
Yep, this is totally on me @folletto 😄

@nosolosw summed it up excellently. I'll just add that having untranspiled code available in `esm/` allows the consuming project to make use of features such as tree-shaking. I.e. with some further iterations, we'll be able to do handy named imports such as

```js
import { GridiconExternal, GridiconGarb } from 'gridicons';
```

while still ensuring that only the code that's relevant for these imports is being pulled in, thanks to tree-shaking -- i.e. we can get both nice named imports, _and_ smaller bundle sizes. This is something we can't do with CommonJS.

(For reference, my original suggestion was at https://github.com/Automattic/gridicons/pull/283#discussion_r172894269)

The different (CJS vs ESM) entrypoints are exposed through the `main` and `esnext` fields in `package.json`; the general idea is taken from http://2ality.com/2017/06/pkg-esnext.html (and as Andrés said, we can use e.g. webpack aliases to still allow 'pretty' import directory filenames, e.g. without the explicit `cjs/` or `esm/` parts).

I've also seen _some_ (albeit not a lot) of standardization towards these names ( `cjs/` or `esm/`), so I'd rather keep them. I'd like to think of them as an implementation detail that typically is aliased by the consumer anyway.

Reply:
Thank you for the reply. The technical answer is grounding for the question I have. I'm concerned on how cryptic now this is. Before "react/build" was clear. Now there's choice, with two three-letter shorthands that require knowing _all of what you described_ to make a choice.

Granted, it's going to be packaged and published on npm, but is there any way to make this more understandable also in this library?

Also: I'm not sure if this is the case but when I read "it would compromise future backward compatibility" are we talking about Calypso or...? If Calypso already works with the more modern format, I'd suggest to ship just that and bump the requirements if needed. Or, at least, create a strong default for that, and leave the other as a secondary option.

At the very very least: can we pick a default, and have a one line explanation on when to use each of these two? Imagine a developer coming in and having no idea on the difference between the two when you write these one liners.

Reply:
'Default' now depends on context: If you project is CommonJS/ES5 only, use `cjs/`. If it supports ES6 stuff -- most notably modules -- either natively, or thru transpilation, use `esm/`. This is somwhat exposed by the relevant fields in `package.json`, but I guess we can also document in the `README`. Sounds good?

Reply:
...which one is Calypso?

Btw, how does that distinction work for people downloading from npm?

Reply:
Well Calypso is transpiled using Babel, so it would be `esm/`. (It'll still work with `cjs/`, which is what our intended upgrade path will be like, to minimize side effects: Merge this, publish to npm, bump Calypso's dependency, and only then switch to `esm/`.)

> Btw, how does that distinction work for people downloading from npm?

Not sure I'm totally getting this. To make things a bit clearer, you could view `esm/` and `cjs/` as basically holding the results from different stages of the build process: `react-to-svg` results in ES6+ files which go to `esm/` (and can be readily used by projects that support those). These files are then transpiled using Babel, and the results end up in `cjs/`. Previously, both types of files would'be ended up in `build/`, but only the transpiled ones would be published to npm. We're essentially now just also including the untranspiled ones. The way we're doing this -- using separate directories and entry points -- mostly reflects constraints from today's JS, and the transitional period we're finding ourselves in, where we cannot yet fully remove transpiled files from an npm, since we might break some consumers which rely on code to be ES5/CommonJS.

Reply:
To sum up, from the consumer point of view we have two alternatives.

The current proposal:

    // live as it was: import the whole iconset in CommonJS format
    import Gridicon from 'gridicons'; 
     <Gridicon icon="external" />

    // Import individual icons in CommonJS format.
    // This is what the README says you should do, 
    // so it'll end up being the option chosen by default.
    import GridiconExternal from 'gridicons/cjs/external';
     <GridiconExternal />

    // Import individual icons in ECMAScript format
    import GridiconExternal from 'gridicons/esm/external';
     <GridiconExternal />

An alternative approach:

    // live as it was: import the whole iconset in CommonJS format
    import Gridicon from 'gridicons'; 
     <Gridicon icon="external" />

    // A nicer path to import CommonJS which is the expected format atm
    import GridiconExternal from 'gridicons/external';
     <GridiconExternal />

    // ECMAScript individual icons
    import GridiconExternal from 'gridicons/esm/external';
     <GridiconExternal />

In addition to what was already said, I'll add that:

* with the alternative, every icon needs to be in the root of the repository. So, after `npm run build` we end up with dozens of files there - to clean that up you need to run a command. This is not the "external API" of the library but it also worth considering the developer experience for Gridicons when talking about this. See this [convo](https://github.com/Automattic/gridicons/pull/283#discussion_r172838047).
* whoever downloads Gridicons with npm to use the React components in their project needs to prepare that code for browsers. It's impossible to use this code without a transpiling tool. Whoever sets up the config, already knows about and needs to be aware of CommonJS / ECMAScript formats.

As I see it, we have a trade-off to make. We either accept breaking backward compatibility in the future when this _module transition_ is over (and go with the alternative now and switch the paths in the future), or we are agnostic about modules (the current proposal). Taking into account that users of the React components already have that knowledge (or copy-paste the README examples, so they're fine) I lean towards the current proposal.

@folletto does it make more sense? would you still prefer the alternative to the current approach?

Reply:
There are several ways how to export stuff from a NPM module:

1. Exporting a CommonJS module: that means using `export.something =` and `module.exports =` statements to export and using `require()` to import dependencies. Code is usually transpiled to ES5. Such a module can be consumed by Node.js directly, and also by all browser bundlers. The oldest and most common format. In `package.json`, the `main` field specifies the location of the module file for this format.

2. Exporting ES6 modules. Code is **still transpiled to ES5**, but imports and exports are done by the respective ES6 statements. These are the only bits of ES6 in an otherwise ES5 code. This module cannot be consumed by Node, and is intended for browser bundlers like Webpack. The bundler will find all the imports and exports, resolves them and outputs an ES5 bundle. No transpilation needs to happen here. This format gradually replaces Common JS for browser libraries. The advantage is that bundlers can optimize ES6 modules much better (tree shaking, marking exports as pure...). In `package.json`, the module file is specified by the `module` field.

3. Exporting untranspiled ES6 modules. Recent versions of Node.js and of major browsers (Chrome, Edge, Firefox, Safari) have full ES6 support, so transpiling is not needed and providing an ES6 version of your module makes sense. That's the `es:next` field and it's very experimental -- there is little mainsteam usage at this moment.

The distinction between option 1 and option 2 is important -- there's been some misconception in this convo that mixes them together.

Gridicons don't need to specify an `es:next` field. What we really want here is `module`.

Reply:
How should the import syntax for Gridicons look like? I believe that the `cjs` and `esm` strings shouldn't be a part of the module's user interface at all. That's an internal detail of the module that Webpack negotiates with its `package.json` and chooses the right option.

Does the ideal interface look like this?
```js
import Gridicon from 'gridicons'; // full package, default import
import { GridiconExternal } from 'gridicons'; // named import
// alternative, less that ideal IMHO
import GridiconExternal from 'gridicons/external'; // import from subpackage
```
or in CommonJS version like this?
```js
var Gridicon = require( 'gridicons' );
var GridiconExternal = require( 'gridicons/external' );
```

Given that we want the bundle to be optimized, i.e., unused icons are not bundled:
- the ES6 ideal above, with one `gridicons` package, is implementable only in Webpack 4 with its `sideEffects` flag.
- the subpackage ES6 import can be implemented without any trouble
- the CommonJS version with one `gridicons` package cannot be efficiently implemented at all
- the CommonJS version with subpackages can be implemented only by having built files in the root directory, with no `build` or `dist` module.

Is it an option to provide the "import single component" feature only for ES6 module consumers? And the CommonJS consumers always getting the whole package? That would simplify things a lot.

Reply:
@jsnajdr thanks for taking the time. I'm trying to distill your comment to some actionable items. Let me know if this makes sense.

> Is it an option to provide the "import single component" feature only for ES6 module consumers? And the CommonJS consumers always getting the whole package?

The user-facing need I'm trying to address is to help Gridicons to ship only the code that the user needs. I also want to make this work for every project that uses Gridicons, independently on the state of any specific bundler/tooling/project/etc.

With that in mind: 

* I consider exporting individual icons in CommonJS format a core part of the proposal, because that's the ubiquitous format at the moment that most tools understand and are capable of working with.
* I've left out named ES6 imports because they don't provide size reduction unless you use a specific set of tools and config. I see that as a nice-to-have. To control scope and I'd like that to be out of this PR.
* I'm open to remove the `esnext` key in the `package.json` if that's a concern. `esm` modules would be still importable without it and I agree that's experimental and not mainstream yet. It makes sense introduce it later when/if the ecosystem has made its mind.
* Although I consider the current proposal more future-proof and with a better developer experience, I won't be opposed to the alternative [I've outlined here](https://github.com/Automattic/gridicons/pull/283#discussion_r173806612). I'm happy to discuss any other option that accomplishes the same goals.

So, @jsnajdr would you agree to land this PR as it is but removing the `esnext` key in the `package.json`?

Reply:
I feel this conversation expanded a little bit — which is great to explore all the details, but I want to make sure we are all on the same page on the why I started this thread: 

1. We are adding a performance feature to import single icons — that's awesome.
2. Yet, this is a icon set. We are introducing two new folder structures, and a new syntax. Are we sure all this complexity is needed for an icon set? (probably yes) We exploded from a single nice `build` folder with react data, to... well all the complexity of the lengthy discussion above.

To be clear: I'm **not** dismissive here – I know that while I can understand the points above I can't argue pros and cons. I just want to stress that we are aiming for a simple approach, intuitive to both maintain and use, and I need your help to review if a simpler approach is possible, or not. I'm totally fine if the answer is "no", I just want to avoid for the answer to be "no" by default just because we haven't tried. :)

Reply:
> `import { GridiconExternal } from 'gridicons'; // named import`

This is definitely my preferred style.

> the ES6 ideal above, with one `gridicons` package, is implementable only in Webpack 4 with its `sideEffects` flag.

Bummer. I honestly wasn't aware of that 🙁  I thought tree-shaking would work in this case with Webpack >= 3 (2?).
(I might not seeing the whole picture, but doesn't the same hold true for any kind of import? In what cases is it required to specify the `sideEffects` flag to actually get tree-shaking?)

Reply:
I clarified a few extra points with Bernie in DM as I know he understand my line of questioning that might have ended up a bit too heavy for this thread. 

So here's an idea to move this forward in a way that should balance all the requirements:

1. Keep `cjs` and `esm` folder structure (thanks for the clarifications on this!)
2. Update the PR to support the preferred syntax (the one in the comment above by Bernie)
3. Update the readme to make the preferred syntax explicit.
4. Update the readme with a one-line explanation of what `cjs` and `esm` folders are for, and how are they used (which includes a note on the alternative, but unsupported, syntax).

Reply:
> `import GridiconExternal from 'gridicons/external'; // import from subpackage`
>
> the subpackage ES6 import can be implemented without any trouble

@jsnajdr But wouldn't that require moving the individual ESM files to the package root (or an alias on the consumer side) as well?

Reply:
For me, consumers are any project that wants Gridicons, not only Calypso. I actually started this out of a particular need for [happychat-client](https://github.com/Automattic/happychat-client/issues/10) so doing this wouldn't be worth for me it I can't use it there.

I'm committed to make this PR work because it addresses a real need. So, here's one bold thing we could do: let's forget completely about the ECMAScript modules. This is precisely what we are doing in Gridicons at the moment and perhaps it's unnecessary to introduce that complexity right now. The [two options above](https://github.com/Automattic/gridicons/pull/283#discussion_r173806612) would become.

- If we use a folder to contain the icons so the devexp is nicer and we have future backward compatibility with `esm` (I use `dist/` instead of `build/` because is more common):

      // Live as it was: import the whole iconset
      import Gridicon from 'gridicons'; 
      <Gridicon icon="external" />

      // Import individual icons
      import GridiconExternal from 'gridicons/dist/external';
      <GridiconExternal />

- If we publish the icons in the root of the repository (nicer paths but will break backward compatibility when we switch Gridicons to ECMAScript modules):  

      // Live as it was: import the whole iconset
      import Gridicon from 'gridicons'; 
      <Gridicon icon="external" />

      // Import individual icons
      import GridiconExternal from 'gridicons/external';
      <GridiconExternal />

To be honest, I'm not the most qualified person to choose between the two because it boils down to how we want to manage the Gridicons library: do we want backward compatibility or not? Being this PR my first contribution I assumed that we wanted that, but it's totally fine if we don't.

@folletto would this simplify things?
Any of the devs consider this as a no-go? @ockham @jsnajdr @gwwar

Reply:
@folletto @ockham wasn't aware of your comments when I posted [this](https://github.com/Automattic/gridicons/pull/283#discussion_r174093348). Does it modifies your thinking @folletto or you still hold on [this](https://github.com/Automattic/gridicons/pull/283#discussion_r174088976)?

Reply:
> But wouldn't that require moving the individual ESM files to the package root (or an alias on the consumer side) as well?
@ockham Good point, that was a mistake on my side. Both CJS and ES subpackages require moving files to package root. I edited my comment accordingly.

Reply:
@nosolosw WFM!

Reply:
> In what cases is it required to specify the sideEffects flag to actually get tree-shaking?)

In case there is a top-level `index.js` that reexports submodules:
```js
export { default as GridiconNotice } from './notice';
export { default as GridiconSearch } from './search';
```
and we import only `GridiconNotice`, then how does Webpack and UglifyJS know they can omit the `GridiconSearch` import? They don't, because importing the `./search` module can have side effects. For example, it can change some global variable or register a DOM listener.

The `sideEffect: false` flag says that no such side effects exists and unused imports can be safely removed.

Reply:
@nosolosw Exporting just the CJS modules and forgetting about ES6 now works for me. The syntax would be:
```js
import Gridicon from 'gridicons';
import GridiconExternal from 'gridicons/dist/external';
```

In future, when Webpack 4 becomes ubiquitous, we can migrate to full ES6:
```js
import { GridiconExternal } from 'gridicons';
```

In the meantime, we could have a Babel transform that lets us write the named import syntax right now and transform it to a `gridicons/dist/${icon}` import. We have something like this for Lodash and Redux Form already in Calypso. But that's optional.

I hope I summarized everything correctly :)